### PR TITLE
🐛 Add missing API type conversions

### LINKED ISSUE #1363: Conversion is not completely implemented for all API types
/kind bug

Hey folks,
in core CAPI we have a tool called conversion-verifier which loops over all types and ensures the controller-runtime conversion interfaces are implemented correctly, i.e. it checks the types have the Hub / ConvertTo / ConvertFrom functions.

The results for CAPV are:
* type "infrastructure.cluster.x-k8s.io/VSphereClusterIdentity" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha4" marked as storage version but it's not convertible, missing Hub() method
* type "infrastructure.cluster.x-k8s.io/VSphereClusterIdentity" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods
* type "infrastructure.cluster.x-k8s.io/VSphereFailureDomain" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha4" marked as storage version but it's not convertible, missing Hub() method
* type "infrastructure.cluster.x-k8s.io/VSphereFailureDomain" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods
* type "infrastructure.cluster.x-k8s.io/VSphereClusterIdentityList" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha4" marked as storage version but it's not convertible, missing Hub() method
* type "infrastructure.cluster.x-k8s.io/VSphereClusterIdentityList" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods
* type "infrastructure.cluster.x-k8s.io/VSphereDeploymentZoneList" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha4" marked as storage version but it's not convertible, missing Hub() method
* type "infrastructure.cluster.x-k8s.io/VSphereDeploymentZoneList" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods
* type "infrastructure.cluster.x-k8s.io/VSphereFailureDomainList" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha4" marked as storage version but it's not convertible, missing Hub() method
* type "infrastructure.cluster.x-k8s.io/VSphereFailureDomainList" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods
* type "infrastructure.cluster.x-k8s.io/VSphereDeploymentZone" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha4" marked as storage version but it's not convertible, missing Hub() method
* type "infrastructure.cluster.x-k8s.io/VSphereDeploymentZone" in package "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods

I think it would be great to:
* implement the missing methods
* integrate the conversion-verifier in CI

The conversion-verifier can be found here:
* Code: https://github.com/kubernetes-sigs/cluster-api/tree/main/hack/tools/conversion-verifier
* Usage: https://github.com/kubernetes-sigs/cluster-api/blob/main/Makefile#L339-L341 (core CAPI runs it as part of the verify job)

I can provide additional details if needed / if you want to implement this issue.

**Environment:**

- Cluster-api-provider-vsphere version: latest
- Kubernetes version: (use `kubectl version`):  -
- OS (e.g. from `/etc/os-release`):  -

### PR DESCRIPTION
**Pull request by aartij17**
**What this PR does / why we need it**:
- adds missing conversions for API types - `VSphereClusterIdentity`, `VSphereClusterTemplates`
- add a new make target to run `conversion verification` tests

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1363 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/tools/go.mod [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
 	github.com/golangci/golangci-lint v1.43.0
-	github.com/onsi/ginkgo v1.16.4
+	github.com/google/gofuzz v1.2.0 // indirect
+	github.com/imdario/mergo v0.3.12 // indirect
+	github.com/mattn/go-runewidth v0.0.13 // indirect
+	github.com/onsi/ginkgo v1.16.5
 	github.com/vmware/govmomi v0.23.1
-	k8s.io/code-generator v0.22.2
-	sigs.k8s.io/controller-tools v0.7.0
+	k8s.io/code-generator v0.23.0-alpha.4
```
Comment (srm09) [inline hack/tools/go.mod:12]: Let's stay with the v0.22 dependencies for now. Consider reverting all the updates here

File: hack/tools/go.mod [hunk 2]
```diff
@@ -1,13 +1,17 @@
...
+	sigs.k8s.io/cluster-api/hack/tools v0.0.0-20211111175208-4cc2fce2111a
```
Comment (srm09) [inline hack/tools/go.mod:13]: Curious as to why would this be needed here. Maybe revert

Comment (srm09) [reply hack/tools/go.mod:13]: The more I think about this, I think it is required to pull in the `conversion-verifier` binary, right?

**Review by srm09 — CHANGES_REQUESTED** [review]
nits mostly. One comment across the files, updated the Copyright year to `2022` for all newly introduced files.

Also, I think you are missing a new target for conversion verifier in the Makefile under `hack/tools` directory.

File: hack/tools/go.mod (hunk 2 above)
Comment (aartij17) [reply hack/tools/go.mod:13]: > The more I think about this, I think it is required to pull in the `conversion-verifier` binary, right?

Yes, that's right!

File: hack/tools/Makefile [hunk 3]
```diff
@@ -82,6 +83,10 @@ kustomize: $(KUSTOMIZE) $(SRCS) ## Build kustomize
 $(KUSTOMIZE): go.mod
 	go build -tags=tools -o $@ sigs.k8s.io/kustomize/kustomize/v4
 
+conversion-verifier: $(CONVERSION_VERIFIER) $(SRCS) ## fetch CAPI's conversion verifier
+$(CONVERSION_VERIFIER): go.mod
```
Comment (aartij17) [inline hack/tools/Makefile:87]: @srm09 - I have added the new make target here to generate the conversion-verifier binary

**Comment by aartij17** [comment]
> Can you create a GH issue to run this as a presubmit job in test-infra? Otherwise looks good to be merged.
> /lgtm
> /approve

Created an issue for this - https://github.com/kubernetes/test-infra/issues/25075

**Pull request merged**